### ISSUE TITLE
Incorrect error when variable has incompatible default during init.

### ISSUE THREAD

**Issue by jbardin** [hashicorp/terraform#21388]
If a root level variable has an incompatible default value, the error during `init` will report that the upgrade tool needs to be run, and does not report the location of the error at all.

For example:

```
variable "x" {
  type = object({
    x = number
  })

  default = {
    x = "y"
  }
}
```

`$ terraform init`

```
Warning: Skipping backend initialization pending configuration upgrade
...
Terraform has installed the required providers to support the configuration
upgrade process. To begin upgrading your configuration, run the following:
    terraform 0.12upgrade
```

Since this incompatible type is the result of a new 0.12 type declaration, running the config tool will only produce a syntax error pointing to the type declaration which isn't valid HCL1 syntax. 

```
Error: Syntax error in configuration file

  on main.tf line 2, in variable "x":
   2:   type = object({
```

Even though terraform in locally initialized, users may be stuck here because the backend has not yet been configured, or simply because the tool is reporting an error and they wouldn't expect `plan` or `apply` to work with an existing error. However the next `plan` operation will return the complete error:

```
Error: Invalid default value for variable

  on main.tf line 6, in variable "x":
   6:   default = {
   7:     x = "y"
   8:   }
```

**Comment by apparentlymart** [comment]
This is, I think, an expected (though definitely unfortunate) consequence of the heuristic nature of that error message. We could potentially adjust the heuristic some more, but I think there will always be cases where it gets it wrong.

The error message there should also be mentioning running `terraform validate` to see the full set of errors. It's implemented that way because if the input _is_ a 0.11-oriented configuration with syntax that 0.12 considers erroneous then the resulting full error output can be overwhelming in extreme cases, and detract from the main call-to-action in the message.

I think probably the main thing for us to do here is to tweak the copy in the error message some more to try to make it clearer that there are two separate paths here and thus a decision to make:

* If _and only if_ you've literally just upgraded from Terraform 0.11 and ran `terraform init` for the first time on your existing configuration, run `terraform 0.12upgrade`.
* In all other situations, run `terraform validate` to see what problems are present.

I had hoped most users wouldn't run into this because they'd hopefully run `terraform init` _before_ making changes to configuration, then run `terraform validate` or `terraform plan` before committing, and thus they'd find and fix the errors using `terraform validate` and `terraform plan` in an already-initialized directory. If they wait to run `terraform init` until after they've made an incorrect change, or if an incorrect change ends up in version control where someone else tries to work with it, then indeed this heuristic can get tripped up and show the "wrong" error message.

**Comment by jbardin** [comment]
Yeah, I also don't think we can return anything more specific in this case. I think adding a suggestion to `validate` the config would help unblock anyone. I expect users trying out the new type system with complex types will get the declarations wrong at some point in a new config, and encounter this on the first `init`.

**Comment by jbardin** [comment]
I admit I missed the `validate` suggestion on the first run through this, which leads me to think the error is too verbose and adding more conditionals in it is only going to make it worse.

Can we look for the existence of complex types `MaybeAlreadyUpgraded` too? It would be nice to skip the upgrade text anyway for users already fully on 0.12 compatible config, since many are going to be experimenting without explicitly setting the `required_version` in each config.

**Comment by apparentlymart** [comment]
"Fully 0.12 compatible" is currently defined such that an error reading with the 0.12 loader means "not compatible". Since the new config loader will still return a partial result for certain errors, we could potentially write some additional logic that sniffs (carefully!) inside that partial result to look for clues that this configuration is intended to already be 0.12 compatible.

Whether the `type` argument would be present in the partial result in this case I'm not sure... since it's the `variable` block whose loading failed, it might be missing or incomplete. That's not to say that we couldn't change the loader to try harder to preserve a partial result here though, if sniffing around in there seems like a compelling approach.

**Comment by jbardin** [comment]
No longer relevant in current releases. Closing.
